src/System.Private.ServiceModel/src/System/ServiceModel/Channels/BufferedStream.cs

```diff
@@ -0,0 +1,1098 @@
+
+        private byte[] _buffer; // Shared read/write buffer.  Alloc on first use.
+
+        private readonly int _bufferSize; // Length of internal buffer (not counting the shadow buffer).
+
+        private int _readPos; // Read pointer within shared buffer.
+        private int _readLen; // Number of bytes read in buffer from _stream.
+        private int _writePos; // Write pointer within shared buffer.
+
+        private Task<int> _lastSyncCompletedReadTask; // The last successful Task returned from ReadAsync
+        // (perf optimization for successive reads of the same size)
+
+        private readonly SemaphoreSlim _sem = new SemaphoreSlim(1, 1);
+
+        public BufferedStream(Stream stream) : this(stream, DefaultBufferSize) {}
+
+
+        public BufferedStream(Stream stream, int bufferSize)
+        {
+            if (stream == null)
+                throw new ArgumentNullException("stream");
+
+            if (bufferSize <= 0)
+                throw new ArgumentOutOfRangeException("bufferSize");
+
+            Contract.EndContractBlock();
+
+            Contract.Assert(!(stream is MemoryStream), "MemoryStream shouldn't be wrapped in a BufferedStream!");
+            Contract.Assert(!(stream is BufferedStream),
+                "BufferedStream shouldn't be wrapped in another BufferedStream!");
+
+            _stream = stream;
+            _bufferSize = bufferSize;
+
+            // Allocate _buffer on its first use - it will not be used if all reads
+            // & writes are greater than or equal to buffer size.
+
+            if (!_stream.CanRead && !_stream.CanWrite)
+                throw new ObjectDisposedException("BufferedStream");
+        }
+
+
+        private void EnsureNotClosed()
+        {
+            if (_stream == null)
+                throw new ObjectDisposedException("BufferedStream");
+        }
+
+
+        private void EnsureCanSeek()
+        {
+            Contract.Requires(_stream != null);
+
+            if (!_stream.CanSeek)
+                throw new NotSupportedException("seek");
+        }
+
+
+        private void EnsureCanRead()
+        {
+            Contract.Requires(_stream != null);
+
+            if (!_stream.CanRead)
+                throw new NotSupportedException("read");
+        }
+
+
+        private void EnsureCanWrite()
+        {
+            Contract.Requires(_stream != null);
+
+            if (!_stream.CanWrite)
+                throw new NotSupportedException("write");
+        }
+
+        /// <summary><code>MaxShadowBufferSize</code> is chosen such that shadow buffers are not allocated on the Large Object Heap.
+        /// Currently, an object is allocated on the LOH if it is larger than 85000 bytes.
+        /// We will go with exactly 80 KBytes, although this is somewhat arbitrary.</summary>
+        private const int MaxShadowBufferSize = 81920; // Make sure not to get to the Large Object Heap.
```

Comment:
There is so much code which has behavior based on the LOH threshold being 85K that I think changing the threshold would be considered a breaking change. It's now .Net dogma :)